src/associations/has-many.js

```diff
@@ -346,16 +346,25 @@ class HasMany extends Association {
       update[this.foreignKey] = null;
 
       updateWhere = {
-        [this.target.primaryKeyAttribute]: obsoleteAssociations.map(associatedObject => associatedObject[this.target.primaryKeyAttribute]),
+        [this.target.primaryKeyAttribute]: obsoleteAssociations.map(
+          associatedObject => associatedObject[this.target.primaryKeyAttribute],
+        ),
       };
 
-      promises.push(this.target.unscoped().update(
-        update,
-        {
-          ...options,
-          where: updateWhere,
-        },
-      ));
+      // update or delete obsolete associations depending on foreign key
+      let udMethod;
+      const udArgs = [{ ...options, where: updateWhere }];
+      const foreignKeyIsNullable
+        = this.target.rawAttributes[this.identifierField].allowNull;
+
+      if (foreignKeyIsNullable) {
+        udMethod = this.target.update.bind(this.target.unscoped());
+        udArgs.unshift(update);
+      } else {
+        udMethod = this.target.destroy.bind(this.target.unscoped());
+      }
```

Comment:
This kind of code will be impossible to type during the TS migration, prefer this instead:

```suggestion
      const target = this.target.unscoped();
      if (foreignKeyIsNullable) {
        promises.push(target.update({
          [this.foreignKey]: null,
        }, {
          ...options,
          where: updateWhere,
        }));
      } else {
        promises.push(target.destroy({
          ...options,
          where: updateWhere,
        }))
      }
```

I think it's more readable too

Reply:
Definitely agree on readability! Also really like the dynamic property change. Didn't think of that, but super clean.